wremnants/muon_calibration.py

```diff
@@ -210,71 +304,144 @@ def define_corrected_muons(df, cvh_helper, jpsi_helper, args, dataset, smearing_
 
     return df
 
-def get_good_gen_muons_idx_in_GenPart(df, reco_subset = "goodMuons"):
-    df = df.Define("goodMuons_idx", f"Muon_genPartIdx[{reco_subset}]")
-    df = df.Define("goodMuonsByGenTruth",
+def getColName_genFiltered_recoMuonSel(reco_sel = "goodMuons", require_prompt = True):
+    genMatch_condition = "Prompt" if require_prompt else "FSonly"
+    return f"{reco_sel}_genTruth_{genMatch_condition}"
+
+def define_genFiltered_recoMuonSel(df, reco_sel = "goodMuons", require_prompt = True):
+    col_name = getColName_genFiltered_recoMuonSel(reco_sel, require_prompt)
+    require_prompt = "true" if require_prompt else "false"
+    df = df.Define(
+        col_name,
         (
-        "ROOT::VecOps::RVec<bool> res(goodMuons_idx.size());"
-        "for (int i = 0; i < goodMuons_idx.size(); i++) {"
-        "    res[i] = ("
-        "        (GenPart_statusFlags[goodMuons_idx[i]] & 0x01) &&"
-        "        (abs(GenPart_pdgId[goodMuons_idx[i]]) == 13) &&"
-        "        (GenPart_status[goodMuons_idx[i]] == 1)"
-        "    );"
-        "}"
-        "return res;"
+            f"wrem::filterRecoMuonsByGenTruth("
+            f"    {reco_sel},"
+             "    Muon_genPartIdx," 
+             "    GenPart_pdgId,"
+             "    GenPart_statusFlags,"
+             "    GenPart_status,"
+            f"    {require_prompt}"
+             ")"
+        )
+    )
+    return df
+
+def define_covMatFiltered_recoMuonSel(df, reco_sel = "goodMuons"):
+    df = df.Redefine(
+        f"{reco_sel}",
+        (
+            f"wrem::filterRecoMuonsByCovMat("
+            f"    {reco_sel},"
+             "    Muon_cvhMomCov_Vals,"
+             "    Muon_cvhMomCov_Counts"
+             ")"
+
         )
     )
-    df = df.Define("goodGenMuons_idx","goodMuons_idx[goodMuonsByGenTruth]")
-    df = df.Filter("goodGenMuons_idx.size() > 0")
     return df
 
 def muon_var_name(mu_type, var):
         return mu_type+(f"_{var}" if mu_type == "Muon" else var.capitalize())
 
-def define_good_gen_muon_kinematics(df, kinematic_vars = ["pt", "eta", "phi", "charge"]):
+def define_matched_gen_muons_covMat(df, reco_sel = "goodMuons"):
+    df = df.Define(
+        f"{reco_sel}_covMat",
+        (
+            "wrem::getCovMatForSelectedRecoMuons<float>("
+            f"    Muon_cvhMomCov_Vals, Muon_cvhMomCov_Counts, {reco_sel}"
+            ");"
+        )
+    )
+    return df
+
+def redefine_matched_gen_muons_covMat(df, reco_sel = "goodMuons"):
+    df = df.Redefine(
```

Comment:
Why do we need redefine here? Is it just to refuse a variable or is it needed?

Reply:
I will remove this function. It was needed because I first defined the covmat column, then used this column to filter gen-matched RECO muons, and then make the same column with the filtered collection of muons. Now I just do the filter by reading directly from the branches in NanoAOD, so no need to redefine covmats.

Reply:
deleted in latest commit